client/src/sdk/GameChannel.ts

```diff
@@ -111,34 +121,70 @@ export class GameChannel {
   }
 
   getStatus() {
-    if (!this.channelInstance) {
-      throw new Error('Channel is not open');
-    }
     return this.getChannelWithoutProxy().status();
   }
 
   async signTx(
     tag: string,
     tx: EncodedData<'tx'>,
-    options?: any
+    options?: any,
+    popupData?: Partial<PopUpData>
   ): Promise<EncodedData<'tx'>> {
+    popupData = popupData ?? {};
     const update = options?.updates?.[0];
+
+    // if we are signing a transaction that updates the contract
     if (update?.op === 'OffChainNewContract') {
       const proposedBytecode = update.code;
-      await verifyContractBytecode(toRaw(this.sdk), proposedBytecode);
+      const isContractValid = await verifyContractBytecode(
+        this.getSdkWithoutProxy(),
+        proposedBytecode
+      );
+      popupData.title = 'Contract validation';
+      popupData.text = `Contract bytecode is 
+      ${isContractValid ? 'matcing' : 'not matching'}`;
+      popupData.mainBtnText = 'Accept Contract';
+      popupData.secBtnText = 'Decline Contract';
+      popupData.mainBtnAction = async () => {
+        await this.buildContract(tx, update.owner);
+      };
+    }
+    if (this.autoSign) {
+      return new Promise((resolve) => {
+        resolve(this.getSdkWithoutProxy().signTransaction(tx, {}));
+        // call the callback if it exists
+        popupData?.mainBtnAction?.();
+      });
     }
-    // TODO show in pop up
-    return new Promise((resolve, reject) => {
-      resolve(Promise.resolve(toRaw(this.sdk).signTransaction(tx, {})));
-      reject(new Error(`Error while signing tx with tag ${tag}`));
+    const popupStore = usePopUpStore();
+    return new Promise((resolve) => {
+      popupStore.showPopUp({
+        title: popupData?.title ?? `Signing ${tag}`,
+        text: popupData?.text ?? 'Do you want to sign this transaction?',
+        mainBtnText: popupData?.mainBtnText ?? 'Confirm',
+        secBtnText: popupData?.secBtnText ?? 'Cancel',
+        mainBtnAction: () => {
+          popupStore.resetPopUp();
+          resolve(this.getSdkWithoutProxy().signTransaction(tx, {}));
+          // call the callback if it exists
+          popupData?.mainBtnAction?.();
+        },
+        secBtnAction: () => {
+          popupStore.resetPopUp();
+          // @ts-expect-error reject the promise
+          resolve(1);
```

Comment:
```suggestion
    return new Promise((resolve, reject) => {
      popupStore.showPopUp({
        title: popupData?.title ?? `Signing ${tag}`,
        text: popupData?.text ?? 'Do you want to sign this transaction?',
        mainBtnText: popupData?.mainBtnText ?? 'Confirm',
        secBtnText: popupData?.secBtnText ?? 'Cancel',
        mainBtnAction: () => {
          popupStore.resetPopUp();
          resolve(this.getSdkWithoutProxy().signTransaction(tx, {}));
          // call the callback if it exists
          popupData?.mainBtnAction?.();
        },
        secBtnAction: () => {
          popupStore.resetPopUp();
          reject(1);
```

Not sure about this one's goal, but if it is to reject, `reject` function could be used instead

Reply:
Even if we don't sign, we need to resolve the promise, there is no error happening, we just want to decline the tx. If we reject we get an error.